source/extensions/filters/http/cache/cache_filter.cc

```diff
@@ -88,7 +88,7 @@ Http::FilterHeadersStatus CacheFilter::encodeHeaders(Http::ResponseHeaderMap& he
     return Http::FilterHeadersStatus::Continue;
   }
 
-  if (filter_state_ == FilterState::ValidatingCachedResponse && isResponseNotModified(headers)) {
+  if (filter_state_ == FilterState::ValidatingCachedResponse && isSuccessfulValidation(headers)) {
```

Comment:
Broadened the scope of the check here, because we want to delete the entry in our cache once validation fails. There isn't an interface in `http_cache` for deleting entries. So we'll pretend it's a validation failure and let an insert to overwrite the content.

Reply:
When we're validating in response to a non-conditional GET request, and the origin returns a 304 that doesn't properly match the cache entry, we still need to serve a 200 from the cache. This change would result in serving the origin's 304. If the response fails the validator check (4.3.4), I think the best we can do is serve the cached entry as-is, then delete it (since we can't update it). Otherwise, we should update the entry, even if the Vary fields have changed.